File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        var res : MutableList<List<Int>> = mutableListOf()

        if(root == null)
            return res

        var queue = ArrayDeque<TreeNode>()
        
        queue.add(root)

        while(queue.size > 0) {
            var currentList : MutableList<Int> = mutableListOf()
            val levelSize = queue.size
            
            for (i in 1..levelSize) {
                val front = queue.first()
                queue.removeFirst()

                if(front.left != null) {
                    queue.addLast(front.left)
                }

                if(front.right != null) {
                    queue.addLast(front.right)
                }

                currentList.add(front.`val`)
            }
            
            res.add(currentList.toList())
        }

        return res
    }   
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int getLengths(TreeNode* root) {
        if(root == nullptr) {
            return 0; 
        }

        int left = 1 + getLengths(root->left); 
        int right = 1 + getLengths(root->right); 

        if(abs(left - right) > 1) {
            return -1; 
        }

        return std::max(left, right);  
    }

    bool isBalanced(TreeNode* root) {
        if(root == nullptr)
            return true; 

        bool l = isBalanced(root->left); 

        int left = getLengths(root->left); 
        int right = getLengths(root->right); 

        if(left == -1 || right == -1)
            return false; 

        if(abs(left - right) > 1)
            return false;

        bool r = isBalanced(root->right);

        if(l != r)
            return false; 

        return true; 
    }
};
